Fixes #1460 - Allow configuring resource roles per app

kolloch opened the pull request:
and provide a possibility to configure a global default.

aquamatthias commented on docs/docs/command-line-flags.md:
```diff
@@ -69,7 +69,15 @@ The core functionality flags can be also set by environment variable `MARATHON_O
-* `--mesos_role` (Optional. Default: None): Mesos role for this framework.
+* `--mesos_role` (Optional. Default: None): Mesos role for this framework. If set, Marathon receives resource offers
+    for the specified role in addition to resources with the role designation '*'.
+* `--default_accepted_resource_roles` (Optional. Default: all roles): Default for the `"defaultAcceptedResourceRoles"`
```
Default for `acceptedResourceRoles`.

kolloch replied:
jepp.

aquamatthias commented on docs/docs/command-line-flags.md:
```diff
@@ -69,7 +69,15 @@ The core functionality flags can be also set by environment variable `MARATHON_O
+* `--default_accepted_resource_roles` (Optional. Default: all roles): Default for the `"defaultAcceptedResourceRoles"`
+    attribute as a comma-separated list of strings. All app definitions which do not specify this attribute explicitly
+    use this value. This defaults to all roles
+    for which this Marathon instance is configured to receive offers (see `--mesos_role`). Changing this configuration
```
I think it is better to describe _when_ this default  applies and what the consequence is. 
I would omit hints regarding restarts in usage info.

kolloch replied:
I asked @pyronicide to help me with the text.

aquamatthias commented on src/main/scala/mesosphere/marathon/tasks/DefaultTaskFactory.scala:
```diff
@@ -21,6 +21,12 @@ class DefaultTaskFactory @Inject() (
   private[this] val log = LoggerFactory.getLogger(getClass)
 
+  // validate here and not in task builder so that we detect the error on startup
+  require(config.defaultAcceptedResourceRolesSet.nonEmpty, "--default_accepted_resource_roles must not be empty")
```
can we do the validation directly in the scallop option?

kolloch replied:
ah cool, didn't know that this is possible. Will do that.

aquamatthias commented on src/main/scala/mesosphere/marathon/tasks/ResourceUtil.scala:
```diff
@@ -109,10 +114,10 @@ object ResourceUtil {
-      usedResourceMap.get(resource.getName) match {
+      usedResourceMap.get(ResourceId(resource)) match {
         case Some(usedResource) =>
           if (resource.getType != usedResource.getType) {
```
Would it make sense to also add the resource type into the ResourceId, so that we don't have to handle that case specifically?

kolloch replied:
Hmmm. it's possible but is it what we want? We wouldn't get warnings for this anymore and I think a warning is appropriate. Don't you?

aquamatthias commented on src/main/scala/mesosphere/mesos/ResourceMatcher.scala:
```diff
@@ -8,22 +8,28 @@ import org.apache.log4j.Logger
+          }.map(_.getRole)
+      }
 
     val cpuRoleOpt = findScalarResourceRole(Resource.CPUS, app.cpus)
```
Can we move from val to def here for all opts? The for comprehension below will only evaluate, if all preceding options are defined.

kolloch replied:
these are very, very cheap, this has nothing to do with the PR but if I can make you happy that easily ;)

kolloch replied:
well the constraint/port matching might not be cheap (even though it is usually in practice). So that might make things better.

aquamatthias commented on src/main/scala/mesosphere/mesos/TaskBuilder.scala:
```diff
@@ -29,20 +29,31 @@ class TaskBuilder(app: AppDefinition,
-        )
-        None
+    if (log.isDebugEnabled) {
+      log.debug(s"acceptedResourceRoles $acceptedResourceRoles")
```
Just a side note: we should change the logging to something different to not test for isDebugEnabled...

kolloch replied:
that's right.

aquamatthias commented on src/main/scala/mesosphere/mesos/TaskBuilder.scala:
```diff
@@ -29,20 +29,31 @@ class TaskBuilder(app: AppDefinition,
+
+    ResourceMatcher.matchResources(
+      offer, app, taskTracker.get(app.id),
+      acceptedResourceRoles = acceptedResourceRoles) match {
```
why a named parameter here?

kolloch replied:
because the type of acceptedResourceRoles is fairly generic (Set[String]) and I like to use named parameters to avoid confusion in these cases.

kolloch commented:
I added quite some tests now and addressed the comments above.

REST API documentation is still missing. I will fix that next.

aquamatthias commented:
Initial review done. 
I had some comments, but overall I agree with this approach.
We need to do some real world testing, before we merge it into master.

kolloch commented:
I added some REST API documentation. I also provided some background information there because I think it is hard to understand otherwise. 

@lloesche helped me to test it. We found one bug which I fixed now (forgot the writing part of the google protocol buffer serialization). Otherwise it looked good.

We used the following test setup:

* Start the Mesos master with `--roles=public`
* Start one Mesos slave with `--ip=127.0.0.1`
* Start another Mesos slave on `--ip=127.0.0.2`, a different work dir and `--default_role=public` 
* Start the Marathon Master with `--mesos_role public --default_accepted_resource_roles \*`

Now we can create a "private" app like this:

```http
PUT /v2/apps/service-private
```
```json
{
  "id": "service-private",
  "cmd": "python -m SimpleHTTPServer $PORT0",
  "cpus": 0.0001,
  "instances": 1,
  "mem": 32,
  "ports": [0]
}
```

and a public app like this

```http
PUT /v2/apps/service-public
```
```json
{
  "id": "service-public",
  "cmd": "python -m SimpleHTTPServer $PORT0",
  "cpus": 0.0001,
  "instances": 1,
  "acceptedResourceRoles": [
    "public"
  ],
  "mem": 32,
  "ports": [ 0]
}
```

Scale up and have fun! :)

pyronicide commented on docs/docs/rest-api.md:
```diff
@@ -195,6 +198,30 @@ Valid constraint operators are one of ["UNIQUE", "CLUSTER",
+##### acceptedResourceRoles
+
+Optional. A list of resource roles. Marathon considers only resource offers with roles in this list for launching
+tasks of this app. If you do not specify this, Marathon considers all resource offers except you override the
```
If you do not specify this, Marathon considers all resource offers that have been configured by the `--default_accepted_resource_roles` command line flag. Note that this defaults to "*" by default.

kolloch replied:
I will try to rephrase this. Note that it does NOT default to "*" if you specify --mesos_role as well.

pyronicide commented:
I had one minor documentation update. This looks like it'll do exactly what we need.

aquamatthias commented on src/main/scala/mesosphere/marathon/MarathonConf.scala:
```diff
@@ -61,9 +61,41 @@ trait MarathonConf extends ScallopConf with ZookeeperConf with IterativeOfferMat
+      "Default for the defaultAcceptedResourceRoles attribute of all app definitions" +
+        " as a comma-separated list of strings." +
+        "This defaults to all roles for which this Marathon instance is configured to receive offers.",
+    default = Some(expectedResourceRoles.mkString(",")),
```
This default value leads to an exception while generating the usage with --help.
```
> run --help
[info] Running mesosphere.marathon.Main --help
[2015-05-18 16:11:59,605] INFO Starting Marathon 0.8.2-SNAPSHOT (mesosphere.marathon.Main$:87)
[error] (run-main-1) org.rogach.scallop.exceptions.UnknownOption: Unknown option 'help'
org.rogach.scallop.exceptions.UnknownOption: Unknown option 'help'
	at org.rogach.scallop.Scallop$$anonfun$5.apply(Scallop.scala:125)
	at org.rogach.scallop.Scallop$$anonfun$5.apply(Scallop.scala:125)
	at scala.Option.getOrElse(Option.scala:120)
	at org.rogach.scallop.Scallop.parse(Scallop.scala:124)
	at org.rogach.scallop.Scallop.parse(Scallop.scala:85)
	at org.rogach.scallop.Scallop.org$rogach$scallop$Scallop$$parsed$lzycompute(Scallop.scala:187)
	at org.rogach.scallop.Scallop.org$rogach$scallop$Scallop$$parsed(Scallop.scala:175)
	at org.rogach.scallop.Scallop$$anonfun$get$5.apply(Scallop.scala:540)
	at org.rogach.scallop.Scallop$$anonfun$get$5.apply(Scallop.scala:537)
	at scala.Option.map(Option.scala:145)
	at org.rogach.scallop.Scallop.get(Scallop.scala:537)
	at org.rogach.scallop.ScallopConf$$anon$1$$anonfun$fn$1.apply(ScallopConf.scala:126)
	at org.rogach.scallop.ScallopConf$$anon$1$$anonfun$fn$1.apply(ScallopConf.scala:126)
	at org.rogach.scallop.ScallopOption.get(ScallopOption.scala:25)
	at mesosphere.marathon.MarathonConf$class.expectedResourceRoles(MarathonConf.scala:69)
	at mesosphere.marathon.MarathonApp$AllConf.expectedResourceRoles(Main.scala:76)
	at mesosphere.marathon.MarathonConf$$anonfun$defaultAcceptedResourceRoles$1.apply(MarathonConf.scala:81)
	at mesosphere.marathon.MarathonConf$$anonfun$defaultAcceptedResourceRoles$1.apply(MarathonConf.scala:81)

```

kolloch replied:
Good that you noticed that.

Fixed now. I also added an integration test for that.

kolloch commented:
I should probably also adjust the JSON schema.

kolloch commented:
Rebased and added JSON schema extension.
